service/api/service.go

```diff
@@ -904,3 +926,62 @@ func (s *serviceImpl) handleError(err error, apiPath string, workflowId string)
 		err.Error(),
 	)
 }
+
+// writeDataObjectsToExternalStorage processes data objects and writes large ones to external storage
+func (s *serviceImpl) writeDataObjectsToExternalStorage(ctx context.Context, dataObjects []iwfidl.KeyValue, workflowId string) ([]iwfidl.KeyValue, error) {
+	if !s.config.ExternalStorage.Enabled {
+		return dataObjects, nil
+	}
+
+	processedDAs := make([]iwfidl.KeyValue, 0, len(dataObjects))
```

Comment:
You don't need to allocate a new slice, I think replacing/processing in the existing slice is more efficient on memory.